tsl/src/compression/api.c

```diff
@@ -773,16 +900,18 @@ tsl_decompress_chunk(PG_FUNCTION_ARGS)
 	if (!ht->fd.compressed_hypertable_id)
 		ereport(ERROR, (errcode(ERRCODE_INTERNAL_ERROR), errmsg("missing compressed hypertable")));
 
-	if (!ts_chunk_is_compressed(uncompressed_chunk))
+	if (ts_relation_uses_hyperstore(uncompressed_chunk_id))
```

Comment:
> So if you run decompress chunk on a hyperstore chunk, it just changes the access method without actually decompressing? You would need to call decompress_chunk twice to decompress it actually?
> 

No, this is not how it works. Calling decompress_chunk() turns the table back to a plain uncompressed heap (in other words, it undo:s what compress_chunk() did), just like the function does for existing compression.  It was agreed during the design phase that we'd keep compatibility with existing APIs and use cases, so it is possible to use compress_chunk() and decompress_chunk() transparently with Hyperstore without learning new APIs.

For example, you'd like to have your most recent chunks as plain heap and then use a compression policy to compress _old_ chunks. With this patch, this use case remains with the difference that compressed chunks can instead be Hyperstores.

For more information, there's a Slab page describing the APIs and semantics.

> Also, a personal nit, but: every time you compress/decompress with hyperstore you need to change the access method of the chunk?
> 
Yes and No. If you are referring to the usage of existing APIs, compress_chunk() and decompress_chunk(), then yes, it will _fully_ compress/decompress and change TAM at the same time (transparently). But these APIs are optional, and are there for compatibility with existing usage and use cases.

Once you are using Hyperstore, there is no need to change access method unless you want to go back to the original heap and fully decompressed.

Also, to elaborate some more, there is no reason that you would actually explicitly want to decompress your whole Hyperstore, because that is essentially the same as having a plain heap. A fully decompressed Hyperstore might still happen implicitly due to DML, but that is probably not desirable and Hyperstore (just like compression) is not optimized for DML. It is also possible to have a fully non-compressed Hyperstore if you start out with an empty Hyperstore. But we are not targeting that use case initially. (Remember that this is just the first iteration of Hyperstore.)

> In my mind i always though hyperstore access method works with fully uncompressed chunks to make things simpler. Having to maintain the access method breaks encapsulation for me. Seems just like maintaining the chunk status.

Not sure what you mean by "maintain the access method".  Hyperstore can certainly have fully non-compressed chunks, but what would be the point? Hyperstore is created to encapsulate compression, not plain heap. If the use case would be to wrap plain (non-compressed) heap, then Hyperstore would just be a worse heap. 

The ideal use case for Hyperstore is to have compressed data. And, while it supports non-compressed data too, the only reason you'd have non-compressed data is due to DML. But non-compressed data is a transitional state until you get back to fully compressed again.

Reply:
I believe we've discussed this and resolved any misunderstandings. Let me know if there are still questions here?